test/metabase/db/custom_migrations_test.clj

```diff
@@ -118,6 +118,57 @@
                      mi/normalize-visualization-settings
                      (#'mi/migrate-viz-settings)))))))))
 
+(deftest add-join-alias-to-visualization-settings-field-refs-test
+  (testing "Migrations v47.00-029: update visualization_settings.column_settings legacy field refs"
+    (impl/test-migrations ["v47.00-029"] [migrate!]
+      (let [result_metadata
+            [{:field_ref [:field 1 nil]}
+             {:field_ref [:field 1 {:join-alias "Self-joined Table"}]}
+             {:field_ref [:field 2 {:source-field 3}]}
+             {:field_ref [:field "column_name" {:base-type :type/Text}]}
+             {:field_ref [:name "column_name"]}]
+            visualization-settings
+            {"column_settings" (-> {["ref" ["field" 1 nil]]                                   {"column_title" "1"}
```

Comment:
This `["ref" ["field" 1 nil]]` column_setting key matches against the first two field_refs in `result_metadata`. All the other `column_setting` keys will only match with one.